src/main/kotlin/com/jetbrains/snakecharm/lang/psi/SMKRuleReference.kt

```diff
@@ -23,5 +22,25 @@ class SMKRuleReference(node: ASTNode): PyElementImpl(node), PsiNamedElement {
         return this
     }
 
+    override fun getReference(): PsiReference? = SMKRulePSIReference(this, TextRange(0, textLength))
+
     private fun getNameNode() = getIdentifierNode(node)
+}
+
+class SMKRulePSIReference(
+        element: PsiElement,
+        textRange: TextRange
+) : PsiReferenceBase<PsiElement>(element, textRange) {
+    private val key: String = element.text
+
+    override fun resolve(): PsiElement? =
```

Comment:
Consider implementing `PsiPolyVariantReference` for multi-resolve (e.g. if several names with same name)

Reply:
If I understand correctly, if there are multiple rules a reference can be resolved to, that means all of those rules are in different files. How should I go about collecting rules from all files in the project? I could do so in a straightforward manner, but I remember something about caching was previously mentioned.

Reply:
Or It could be several rules with the same name in the same file. Such case will be highlighted as error, but resolve could work in any case. I suggest implement it without collecting references from the same file, e.g get rules from SmkRule object.

Reply:
>e.g get rules from SmkRule object

I don't understand what this means. Did you mean `SnakemakeFile`? How do I get rules from `SmkRule`?

So, I have to get the project from the `SnakemakeFile` that contains the current element, go over all .smk files in that project path, and collect references from each of those files except for the current one, is that what you meant?

Reply:
> Did you mean SnakemakeFile

Yes

> go over all .smk files in that project path

No, just same SMK file at the moment.